pilot/pkg/xds/ads.go

```diff
@@ -623,10 +617,28 @@ func setTopologyLabels(proxy *model.Proxy) {
 			SubZone: proxy.XdsNode.Locality.GetSubZone(),
 		}
 	}
-
-	locality := util.LocalityToString(proxy.Locality)
 	// add topology labels to proxy labels
-	proxy.Labels = labelutil.AugmentLabels(proxy.Labels, proxy.Metadata.ClusterID, locality, proxy.GetNodeName(), proxy.Metadata.Network)
+	proxy.Labels = labelutil.AugmentLabels(proxy.Labels, proxy.Metadata.ClusterID, util.LocalityToString(proxy.Locality), proxy.GetNodeName(), proxy.Metadata.Network)
+}
+
+func localityFromProxyLabels(proxy *model.Proxy) *core.Locality {
+	region, f1 := proxy.Labels[labelutil.LabelTopologyRegion]
+	zone, f2 := proxy.Labels[labelutil.LabelTopologyZone]
+	subzone, f3 := proxy.Labels[label.TopologySubzone.Name]
+	if !f1 && !f2 && !f3 {
+		// If no labels set, we didn't find the locality from the service registry. We do support a (mostly undocumented/internal)
+		// label to override the locality, so respect that here as well.
+		ls, f := proxy.Labels[model.LocalityLabel]
+		if f {
+			return util.ConvertLocality(ls)
+		}
+		return nil
```

Comment:
Is this necessary now as you have called `AugmentLabels`

Reply:
Yes, we still need to be able to get it from the XdsNode on L615 if the labels are not set